k8s/sonarqube/README.md

```diff
@@ -0,0 +1,462 @@
+This shell script will create `postgresql-backup.sql` dump of all DB in PostgreSQL.
+
+```shell
+kubectl --namespace $NAMESPACE exec -t \
+	$(kubectl -n$NAMESPACE get pod -oname | \
+		sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
+	-c postgresql-server \
+	-- pg_dumpall -c -U postgres > backup.sql
+```
+
+## Restoring
+
+Entering to maintenance mode:
+In order to restore PostgresSQL database there is a need to perform some preliminary steps:
+The command below will “lock” the database by blocking new incoming connections:
+
+```shell
+kubectl --namespace $NAMESPACE exec -t \
+  $(kubectl -n$NAMESPACE get pod -oname | \
+     sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
+  -c postgresql-server \
+  -- psql -U postgres -c "REVOKE CONNECT ON DATABASE sonar FROM PUBLIC;"
+```
+
+Next in order to ensure data consistency all active connections will be dropped:
+
+```shell
+kubectl --namespace $NAMESPACE exec -t \
+  $(kubectl -n$NAMESPACE get pod -oname | \
+     sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
+  -c postgresql-server \
+  -- psql -U postgres -c "select pg_terminate_backend(pid) from pg_stat_activity where datname='sonar';"
+```
+
+Now you are able to restore data to the database.
+Below shell script will restore data from `backup.sql` to PostgreSQL
+
+```shell
+cat backup.sql | kubectl --namespace $NAMESPACE exec -i \
+  $(kubectl -n$NAMESPACE get pod -oname | \
+    sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
+  -c postgresql-server \
+  -- psql -U postgres
+```
+
+Next copy files from current folder to folder `data` in SonarQube application pod:
+
+```shell
+kubectl --namespace $NAMESPACE cp ./ $(kubectl -n$NAMESPACE get pod -oname | \
+  sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/data
+```
+
+Afterwards copy files from current folder to `extensions` folder in SonarQube application pod
+
+```shell
+kubectl --namespace $NAMESPACE cp ./ $(kubectl -n$NAMESPACE get pod -oname | \
+  sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/extensions
+```
+
+Delete all outstanding and unneeded SonarQube application data:
+
+```shell
+kubectl --namespace $NAMESPACE exec -i \
+  $(kubectl -n$NAMESPACE get pod -oname | \
+    sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p) \
+  -- bash -c "rm -rf /opt/sonarqube/data/es5/* "
+```
+
+Exit maintenance mode by unlocking schema `sonar` for new incoming connection:
+
+```shell
+kubectl --namespace $NAMESPACE exec -t \
+  $(kubectl -n$NAMESPACE get pod -oname | \
+    sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
+  -c postgresql-server \
+  -- psql -U postgres -c "GRANT CONNECT ON DATABASE sonar TO PUBLIC;"
+```
+
+Finally restart SonarQube application pod:
+
+```shell
+kubectl --namespace $NAMESPACE delete pod $(kubectl -n$NAMESPACE get pod -oname | \
+  sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p)
+```
+
```

Comment:
After following this procedure, I was unable to connect to SonarQube, with following error on port-forward
```
E0429 16:02:32.162420  260600 portforward.go:331] an error occurred forwarding 9000 -> 9000: error forwarding port 9000 to pod 616d364008da583808ae0a8722899210409ea018f8d3ff2de55ce5c45d4ec879, uid : exit status 1: 2019/04/29 14:02:32 socat[3797717] E connect(5, AF=2 127.0.0.1:9000, 16): Connection refused
```

Reply:
it is normal behavior you dint need port-forwarding for backup-restore procedure